include/real/const_precision_iterator.hpp

```diff
@@ -0,0 +1,391 @@
+#ifndef BOOST_CONST_PRECISION_ITERATOR_HPP
+#define BOOST_CONST_PRECISION_ITERATOR_HPP
+
+#include "real/boundary_helper.hpp"
+#include "real/interval.hpp"
+#include "real/real_explicit.hpp"
+#include "real/real_algorithm.hpp"
+#include "real/real_operation.hpp"
+#include "real/real_helpers.hpp"
+#include <memory>
+#include <variant>
+#include <assert.h>
+#include <iterator>
+
+namespace boost {
+    namespace real{
+    /**
+     * @file The const_precision_iterator provides the functionality to iterate through precision intervals
+     * of all three kinds of reals
+     * 
+     * @note variant and visit/visitors are used extensively in this implementation
+     * @TODO: consider doing traversals to lower the demands of recursion
+     * @sa documention on std::variant, std::visit
+     */
+        template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
+        template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
+        // fwd decl
+        class real;
+        class operation_helpers;
+
+        class const_precision_iterator {
+            public:
+            /**
+             * @brief Determines the maximum precision to use
+             */
+            static std::optional<unsigned int> maximum_precision;
+
+            /// @TODO look into STL-style iterators
+            // typedef std::forward_iterator_tag iterator_category;
+            // typedef void difference_type (?);
+            // typedef ??? value_type
+            // typedef const value_type& reference (?) not necessary because const
+            // typedef const value_type* pointer
+
+            private:
+                typedef std::variant<std::monostate, real_explicit, real_algorithm, real_operation> real_number;
+
+                // needed as friend so that we can update operands' precision itrs
```

Comment:
The _lhs and _rhs of real_operation are private. I can replace accessing them directly with some getter function. Still, doing ++ to a real_operation itr does ++ to its operands' itrs. So, those objects do get modified. I don't see a good way of decoupling this. Though, technically, it is still "const", since the real_operation's values (addresses of operands) do not change (but the values of the operands' precision iterators do change).